sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -545,4 +630,289 @@ public void output(OutputT output, Instant timestamp, BoundedWindow window) {
           WindowedValue.of(output, timestamp, window, PaneInfo.NO_FIRING));
     }
   }
+
+  /**
+   * A {@link StateBinder} that uses the Beam Fn State API to read and write user state.
+   *
+   * <p>TODO: Add support for {@link #bindMap} and {@link #bindSet}. Note that
+   * {@link #bindWatermark} should never be implemented.
+   */
+  private class BeamFnStateBinder implements StateBinder {
```

Comment:
Wondering if its better to move this class to its own file and making package private, so that it would be easier if we wanted to reuse it or add its own tests. Also improves readability of the FnApiDoFnRunner class.

Reply:
That can happen once there is a need for it.